### PR TITLE
Changes to Definition AST nodes for DAO usage

### PR DESCRIPTION
**Pull request by m-cheung**
This pull request prepares for upcoming changes with regards to canonicalization of ASTs. ID fields are added to definitions so they can be identified by them.

### CODE REVIEW

File: lib/org/chromium/webidl/Parser.js [hunk 1]
```diff
@@ -423,7 +423,7 @@ foam.CLASS({
         Definitions: parser.extAttrAndValue(parser.Definition, 'definition'),
         Callback: function(v) {
           if (parser.Callback.isInstance(v)) return v;
-          return parser.CallbackInterface.create({interface: v});
+          return parser.CallbackInterface.create({interfaceDetail: v});
```
Comment (m-cheung) [inline lib/org/chromium/webidl/Parser.js]: Had to change this due to issue described below

File: lib/org/chromium/webidl/ast/CallbackInterface.js [hunk 2]
```diff
@@ -12,14 +12,20 @@ foam.CLASS({
     {
       class: 'FObjectProperty',
       of: 'org.chromium.webidl.ast.Interface',
-      name: 'interface',
+      name: 'interfaceDetail',
     },
+    {
+      name: 'id',
+      expression: function(interfaceDetail) {
+        return interfaceDetail.id;
```
Comment (m-cheung) [inline lib/org/chromium/webidl/ast/CallbackInterface.js]: `interface` cannot be used as a variable as we are running in strict mode and `interface` is a reserved word.
We could just do `ids: [ 'interface' ]` but I'm not certain if comparing interface objects would be efficient

File: lib/org/chromium/webidl/ast/Implements.js [hunk 3]
```diff
@@ -8,6 +8,8 @@ foam.CLASS({
   name: 'Implements',
   extends: 'org.chromium.webidl.ast.DefinitionData',
 
+  ids: [ 'implementer', 'implemented' ],
```
Comment (m-cheung) [inline lib/org/chromium/webidl/ast/Implements.js]: This comment applies to all `ids` fields. Is there any performance concerns that we have to worry about? `implementer` and `implemented` are objects (with a `literal` property that is a string), and I'm not certain if the usage of objects as ID will be optimal.

File: lib/org/chromium/webidl/ast/Named.js [hunk 4]
```diff
@@ -7,6 +7,11 @@ foam.CLASS({
   package: 'org.chromium.webidl.ast',
   name: 'Named',
 
+  // The actual literal is stored in name.literal... will this be an issue?
+  // Also, Operations nodes are Named... but they have optional identifiers...
+  // Thus, name === null for some nodes...
+  ids: ['name'],
```
Comment (m-cheung) [inline lib/org/chromium/webidl/ast/Named.js]: `Operation` nodes implements `Named`, however, they have optional identifiers (as specified by the grammar). In these cases, `name === null` and issues may arise in the future with regards to this.

Comment (mdittmer) [reply lib/org/chromium/webidl/ast/Named.js]: We should only need identity for top-level entities, right?

File: lib/org/chromium/webidl/ast/Definition.js [hunk 5]
```diff
@@ -14,6 +14,10 @@ foam.CLASS({
       of: 'org.chromium.webidl.ast.DefinitionData',
       name: 'definition',
     },
+    {
+      name: 'id',
+      factory: function() { return this.definition.id; },
```
Comment (mdittmer) [inline lib/org/chromium/webidl/ast/Definition.js]: Here, let's delegate to this.getName(), which delegates to definition.getName(), and so on. That way other AST nodes can implement getName(), without needing to have an id.

File: lib/org/chromium/webidl/ast/Implements.js (hunk 3 above)
Comment (mdittmer) [reply lib/org/chromium/webidl/ast/Implements.js]: This is fine for now. Add a TODO that poses the question of whether we should have a custom id property that delegates to implementer and implemented's getName()

File: lib/org/chromium/webidl/ast/Operation.js [hunk 6]
```diff
@@ -22,6 +22,13 @@ foam.CLASS({
   ],
 
   methods: [
+    function getId() {
+      // Constructor-like declarations (e.g. serializer Example()) do not have
+      // names. Instead, it is set as type. Similarly, declarations with
+      // qualifiers (e.g. getter, setter) do not have names either. However,
+      // since IDs are not currently used at the Operation level, we return null
+      return this.name ? this.name.literal : null;
```
Comment (mdittmer) [inline lib/org/chromium/webidl/ast/Operation.js]: We shouldn't need this once we have getName on Named, right?

**Review by mdittmer — CHANGES_REQUESTED** [review]

File: lib/org/chromium/webidl/ast/Definition.js (hunk 5 above)
Comment (m-cheung) [reply lib/org/chromium/webidl/ast/Definition.js]: Not all definitions (maybe it is just `implements`) are named. Perhaps the `ids` field should be removed from `Implements` and replaced with `getName()` which returns those two objects?

**Comment by m-cheung** [comment]
@mdittmer PTAL when you have the chance. I believe all requested changes have been made

File: lib/org/chromium/webidl/ast/Named.js [hunk 7]
```diff
@@ -14,4 +14,10 @@ foam.CLASS({
       name: 'name',
     },
   ],
+
+  methods: [
+    function getName() {
+      return this.name.literal;
```
Comment (mdittmer) [inline lib/org/chromium/webidl/ast/Named.js]: If we want to be more defensive here, we could:

(1) Return `this.name ? this.name.literal || '' : ''`
(2) Implement `validate()` on `Definition` with a call to `this.SUPER()` and a `foam.assert()`ion that `id` is truthy
(3) Call `.validate()` before sending ASTs to a box or putting them to a DAO

**Review by mdittmer — APPROVED** [review]
LGTM. One last suggestion; please add TODO or implement defensive programming suggestion. Then, merge it!

**Pull request merged**